test/misc/queue.js

```diff
@@ -0,0 +1,33 @@
+const assert = require('assert');
+const rewire = require('rewire');
+const { assertIncludes, loader } = require('../utils.js');
+const rollup = rewire('../../dist/shared/rollup');
```

Comment:
I'm not sure we need this, also there seem to be some weird breakages in unrelated tests. And the test is also tied to the implementation. Instead, I would just manually try to patch `fs` it this works using the `before` and `after` hooks of the test. Look at the `function/default-on-warn` test for where to add these hooks.

For Rollup to be able to pick up the patch, you will need to change the import in `utils/fs.ts` from `import * as fs from 'fs';` to `import fs from 'fs';`, otherwise you will not get the shared commonjs fs object but one wrapped by Rollup.

Basically, your max-parallel-file-reads function test could look like this:

```js
const assert = require('assert');
const fs = require('fs');
const fsReadFile = fs.readFile;
let currentReads = 0;
let maxReads = 0;

module.exports = {
	description: 'unlimited parallel file reads',
 	options: {
 		maxParallelFileReads: 3
 	},
	before() {
		fs.readFile = (path, options, callback) => {
			currentReads++;
			maxReads = Math.max(maxReads, currentReads);
			fsReadFile(path, options, (err, data) => {
				currentReads--;
				callback(err, data);
			});
		};
	},
	after() {
		fs.readFile = fsReadFile;
		assert.strictEqual(
 			maxParallelFileReads,
 			3,
 			'Wrong number of parallel file reads: ' + maxParallelFileReads
 		)
	}
};
```

Then you can observe everything strictly from the outside, do not need to track this in Rollup and also do not need to patch into implementation details.

With this setup, you could also easily simulate read errors in a second test as you have the callback in your hand.

Reply:
I actually tried hooking into fs.readFile first but could not get it to work. Your hint with replacing `* as fs` was exactly what I was missing. The tests do make more sense this way, thanks.

I have now replaced the previous tests.